### PR TITLE
Handle nan and infinity in F# analysis results

### LINKED ISSUE #4004: Program doesn't load with blazor
I suspect it's to do with serialization of NaN and Infinity (they're probably serialized at `"NaN"` vs `NaN` in the old one).

```
{
  a : 0.0 / 0.0
  b : 1.0 / 0.0
  c : 5.6
}
```

### PR DESCRIPTION
**Pull request by StachuDotNet**
Resolves #4004 

Concerning the client changes:
Since `Js.Json.decodeObject(j)` fails with these values, I decided to handle these in the `None` branch, which works well enough - does that make sense?

### CODE REVIEW

**Comment by StachuDotNet** [comment]
Should any tests be written around this? I suppose some integration test could be?

**Comment by pbiggar** [comment]
> Should any tests be written around this? I suppose some integration test could be?

Maybe add a decoder unit test in the client?

**Comment by pbiggar** [comment]
> Resolves #4004
> 
> Concerning the client changes: Since `Js.Json.decodeObject(j)` fails with these values, I decided to handle these in the `None` branch, which works well enough - does that make sense?

Yeah, i think that's a good option.

File: client/src/core/Decoders.res [hunk 1]
```diff
@@ -346,6 +346,10 @@ let rec dval = (j): dval => {
      * analysis code.  However, since JSON doesn't support Infinity/NaN, our
      * JSON parser crashes. So instead we encode them specially in
      * Jsanalysis.clean_yojson, which is the encoder which matches this decoder
+     *
+     * CLEANUP remove OCaml-specific parsing here.
+     * The checks within `Some` are for OCaml, while the ones under `None` are
```
Comment (pbiggar) [inline client/src/core/Decoders.res:351]:
```suggestion
     * The checks within `Some` are for OCaml analysis, while the ones under `None` are
```

**Review by pbiggar — APPROVED** [review]

**Comment by pbiggar** [comment]
Feel free to merge with these changes.

**Comment by StachuDotNet** [comment]
I added a test which seems like it should pass, but it fails.

It errors with:
```
Expected: {"TAG":0,"_0":{"k":"1818290858","v":{"TAG":0,"_0":{"TAG":6,"_0":{"k":"b","v":{"TAG":1,"_0":null},"h":2,"l":{"k":"a","v":{"TAG":1,"_0":null},"h":1},"r":{"k":"c","v":{"TAG":1,"_0":5.6},"h":1}}}},"h":1}}
      Actual: {"TAG":0,"_0":{"k":"1818290858","v":{"TAG":0,"_0":{"TAG":6,"_0":{"k":"b","v":{"TAG":1,"_0":null},"h":2,"l":{"k":"a","v":{"TAG":1,"_0":null},"h":1},"r":{"k":"c","v":{"TAG":1,"_0":5.6},"h":1}}}},"h":1}}
```

(the same strings)

I think I'm probably supposed to convert these into... more _sane_ types, but not sure what/how.

@pbiggar Any ideas?

**Comment by StachuDotNet** [comment]
Also, surprised (disappointed) that ReScript isn't auto-formatted via the script

**Comment by pbiggar** [comment]
You can't do float equality like that. NaNs don't equal anything, not even other NaNs.

Just realized you need to handle -Infinity as well.

**Comment by StachuDotNet** [comment]
So, should I just do the test with Infinity and -Infinity?

**Comment by StachuDotNet** [comment]
I'll try to wrangle some IsNan in a test

**Comment by StachuDotNet** [comment]
FYI figured out how to simplify tests - had to study code a bit more.

**Comment by StachuDotNet** [comment]
Although you said to merge when tests are OK, I'd still appreciate a quick review.

I'm pretty happy with this PR - ReScript's kinda nice. Really looking forward to studying/improving `client` over time.

**Comment by pbiggar** [comment]
Nice! You're good to merge.

**Review by pbiggar — APPROVED** [review]

**Pull request merged**